docs/guide/spa-client-configuration.md
# Configuration

## Overview

You can set config file path by `config-dir` commandline option or by environment variables `SPA_CLIENT_CONFIG`,
you can also set all config by environment variables like
[.env](https://github.com/fornetcode/spa-server/blob/master/example/js-app-example/.env). Config override order is
command line option > config file > environment.

## Reference

## Toml Format Config

```toml
[server]
address = "http://127.0.0.1:9000"
auth_token = "token"
# [upload]
## default value is:3
# parallel = 3
```

docs/guide/spa-server-configuration.md
# Configuration

## Overview

The config format toml.

The config default path is './config.conf', you can change it by environment `SPA_CONFIG`.

## Config Reference

## Toml Format Config

```toml
## directory to store static web files. if you use docker, please mount a persistence volume for it.
file_dir = "/data"

## enable cors, default is false, its implementation is simple now.
## Access-Control-Allow-Origin: $ORIGIN
## Access-Control-Allow-Methods: OPTION,GET,HEAD
## Access-Control-Max-Age: 3600
cors = false
## http bind, if set port <= 0 or remove http, will disable http server(need set https config)
[http]
port = 80
addr = "0.0.0.0"
## port when serving public PI,default is http port. external_port should not be 0.
# external_port = 80

## optional, when https enabled, redirect_https default value true
## it would the port would be https.external_port(https.external_port should be defined), otherwise is false
# redirect_https = true

# [https]
# port = 443 # https bind address
# addr = "0.0.0.0"
## port when serving public PI,default is https port. external_port should not be 0.
# external_port = 443

## if set true, http server(80) will send client
## status code:301(Moved Permanently) to tell client redirect to https

## default value for https ssl
# [https.ssl]
# private = "private.key path" # private ssl key
# public = "public.cert path" # public ssl cert

## acme config, it doest not support run with https.ssl config.
# [https.acme]
## emails to Let's Encrypt needs to interact.
# emails = ["mailto:email@example.com"]

## directory to store account and certificate
## optional, default is ${file_dir}/acme
# dir = "/data/acme"

## ci / stage / prod, default is prod, ci is just for CI test with Pebble, don't use it.
# type = prod

## default cache config
[cache]
## if file size > max_size, it will not be cached. default is (10MB).
# max_size = 10_000_000
## gzip compression for js/json/icon/json, default is false,
## only support gzip algo, and only compress cached files,
## be careful to set it true
# compression = false

## http header Cache-Control config,
## optional, if not set, won't sender this header to client
# [[cache.client_cache]]
## 30day
# expire = '30d'
# extension_names = ['icon', 'gif', 'jpg', 'jpeg', 'png', 'js']
# [[cache.client_cache]]
# expire = '0'
# extension_names = ['html']

## admin server config
## admin server don't support hot reload. the config should not change.
## optional, and it's disabled by default.
## if you use spa-client to upload files, control version. Need to open it
# [admin_config]
# port = 9000
# addr = "127.0.0.1"

## this is used to check client request
## put it in http header,  Authorization: Bearer $token
# token = "token"
## max file size allowed to be uploaded,
## default is 30MB(30*1000*1000)
# max_upload_size = 30_1000_1000
## delete deprecated version by cron
# [admin_config.deprecated_version_delete]
## default value: every day at 3am.
# cron = "0 0 3 * * *"
## default value is 2
# max_preserve = 2

## optional, domains specfic config, it will use the default config if not set
# [[domains]]
# domain = "www.example.com"
## optional, `example.com` would redirect to `www.example.com`
# alias = ["example.com"]
# cors = false
# [domains.https]
## optional, when https enabled, redirect_https default value true
## it would the port would be https.external_port(https.external_port should be defined), otherwise is false
# redirect_https = 443

## this would be usefully when set https.acme
# disable_acme = false
# [domains.https.ssl]

# [domains.cache]
# max_size = 10_000_000
# compression = false
# [[domains.cache.client_cache]]
# expire = '30d' # 30day
# extension_names = ['icon', 'gif', 'jpg', 'jpeg', 'png', 'js']


# [openTelemetry]
# endpoint = "http://localhost:4317"
```


server/src/config.rs
use anyhow::{bail, Context};
use duration_str::deserialize_duration;
use serde::Deserialize;
use small_acme::LetsEncrypt;
use std::time::Duration;
use std::{env, fs};
use tracing::warn;

const CONFIG_PATH: &str = "/config/config.toml";

#[derive(Deserialize, Debug, Clone, PartialEq)]
pub struct Config {
    pub file_dir: String,
    #[serde(default)]
    pub cors: bool,
    pub admin_config: Option<AdminConfig>,
    pub http: Option<HttpConfig>,
    pub https: Option<HttpsConfig>,
    #[serde(default)]
    pub cache: CacheConfig,
    #[serde(default)]
    pub domains: Vec<DomainConfig>,
    pub open_telemetry: Option<OpenTelemetry>,
}

#[derive(Deserialize, Debug, Clone, PartialEq)]
pub struct OpenTelemetry {
    pub endpoint: String,
}

//TODO: create config with lots of default value
impl Config {
    pub fn load() -> anyhow::Result<Self> {
        let config_path = env::var("SPA_CONFIG").unwrap_or(CONFIG_PATH.to_string());

        let config = Self::load_toml(&config_path)?;
        if config.http.is_none() && config.https.is_none() {
            bail!("should set http or https server config")
        }
        if let Some(http_config) = &config.https {
            if http_config.acme.is_some() && http_config.ssl.is_some() {
                bail!("spa-server don't support ssl and acme config in the meantime");
            }
            if http_config
                .acme
                .as_ref()
                .is_some_and(|c| c.emails.is_empty())
            {
                bail!("acme emails must be set")
            }
            if http_config
                .acme
                .as_ref()
                .is_some_and(|c| matches!(c.acme_type, ACMEType::CI) && c.ci_ca_path.is_none())
            {
                bail!("acme CI must set ca path")
            }
            if http_config.acme.is_some() && config.http.as_ref().filter(|v| v.port != 80).is_none()
            {
                warn!("acme needs http port:80 to signed https certificate");
            }
        }
        if config
            .domains
            .iter()
            .any(|x| !get_host_path_from_domain(&x.domain).1.is_empty())
        {
            bail!("domains.domain do not support sub path like 'www.example.com/abc' now")
        }
        Ok(config)
    }

    fn load_toml(path: &str) -> anyhow::Result<Self> {
        let config = fs::read_to_string(path)
            .with_context(|| format!("can not read config file: {path}"))?;
        let config: Config = toml::from_str(&config).with_context(|| "parse config file error")?;
        Ok(config)
    }
}
#[derive(Deserialize, Debug, Clone, Eq, PartialEq)]
pub struct AdminConfig {
    pub port: u32,
    pub addr: String,
    pub token: String,
    #[serde(default = "default_max_upload_size")]
    pub max_upload_size: u64,
    pub deprecated_version_delete: Option<DeprecatedVersionRemove>,
}

fn default_max_upload_size() -> u64 {
    30 * 1024 * 1024
}

#[derive(Deserialize, Debug, Clone, PartialEq)]
pub struct DomainConfig {
    pub domain: String,
    pub cors: Option<bool>,
    pub cache: Option<DomainCacheConfig>,
    pub https: Option<DomainHttpsConfig>,
    pub alias: Option<Vec<String>>,
    pub redirect_https: Option<bool>,
}

#[derive(Deserialize, Debug, Clone, PartialEq)]
pub struct DomainHttpsConfig {
    pub ssl: Option<SSL>,
    #[serde(default)]
    pub disable_acme: bool,
}

#[derive(Deserialize, Debug, Clone, PartialEq)]
pub struct SSL {
    pub private: String,
    pub public: String,
}

#[derive(Clone, Debug, Deserialize, Default, PartialEq)]
#[serde(rename_all = "lowercase")]
pub enum ACMEType {
    CI,
    #[default]
    Prod,
    Stage,
}

impl ACMEType {
    pub fn url(&self) -> &'static str {
        match self {
            ACMEType::Stage => LetsEncrypt::Staging.url(),
            ACMEType::Prod => LetsEncrypt::Production.url(),
            ACMEType::CI => "https://localhost:14000/dir",
        }
    }
    //acme_async regex use this.
    pub fn as_str(&self) -> &'static str {
        match self {
            ACMEType::Stage => "stage",
            ACMEType::Prod => "prod",
            ACMEType::CI => "ci",
        }
    }
}

#[derive(Deserialize, Debug, Clone, Default, PartialEq)]
pub struct ACMEConfig {
    pub emails: Vec<String>,
    pub dir: Option<String>,
    #[serde(default, rename = "type")]
    pub acme_type: ACMEType,
    pub ci_ca_path: Option<String>, // this is for CI
}

#[derive(Deserialize, Debug, Clone, PartialEq)]
pub struct HttpConfig {
    pub addr: String,
    pub port: u16,
    pub external_port: Option<u16>,
    pub redirect_https: Option<bool>,
}

#[derive(Deserialize, Debug, Clone, PartialEq)]
pub struct HttpsConfig {
    pub ssl: Option<SSL>,
    pub acme: Option<ACMEConfig>,
    pub port: u16,
    pub external_port: Option<u16>,
    pub addr: String,
    #[serde(default)]
    pub http_redirect_to_https: u16,
}
// should write Deserialize by hand.
#[derive(Deserialize, Debug, Clone, PartialEq)]
pub struct CacheConfig {
    #[serde(default = "default_max_size")]
    pub max_size: u64,
    #[serde(default)]
    pub compression: bool,
    #[serde(default)]
    pub client_cache: Vec<ClientCacheItem>,
}
#[derive(Deserialize, Debug, Clone, PartialEq)]
pub struct DomainCacheConfig {
    pub max_size: Option<u64>,
    pub compression: Option<bool>,
    pub client_cache: Option<Vec<ClientCacheItem>>,
}

fn default_max_size() -> u64 {
    10 * 1000 * 1000
}

#[derive(Deserialize, Debug, Clone, PartialEq)]
pub struct ClientCacheItem {
    #[serde(deserialize_with = "deserialize_duration")]
    pub expire: Duration,
    pub extension_names: Vec<String>,
}

impl Default for CacheConfig {
    fn default() -> Self {
        CacheConfig {
            max_size: default_max_size(),
            client_cache: Vec::new(),
            compression: false,
        }
    }
}
#[derive(Deserialize, Debug, Clone, Eq, PartialEq)]
pub struct DeprecatedVersionRemove {
    #[serde(default = "default_cron")]
    pub cron: String,
    #[serde(default = "default_max_reserve")]
    pub max_reserve: u32,
}
pub fn default_cron() -> String {
    String::from("0 0 3 * * *")
}
pub fn default_max_reserve() -> u32 {
    2
}

pub fn get_host_path_from_domain(domain: &str) -> (&str, &str) {
    match domain.split_once('/') {
        None => (domain, ""),
        Some(v) => v,
    }
}

#[cfg(test)]
mod test {
    use std::env;
    use std::path::PathBuf;

    fn get_project_path() -> PathBuf {
        let path = PathBuf::from(env!("CARGO_MANIFEST_DIR"));
        let path = path.parent().unwrap();
        path.to_owned()
    }
}
